### PR TITLE
Templating resources

### PR DESCRIPTION
**Pull request by danielepolencic**
Preview: https://deploy-preview-275--learnk8s.netlify.com/templating-yaml-with-code

### CODE REVIEW

File: src/templating/content.md [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+
+Kustomize is more structured that `yq` but this convenience comes with some drawbacks.
+
+Kustomize is designed on purpose to [make it hard to change values through the command line](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/eschewedFeatures.md#unstructured-edits).
+
+You have to write YAML files to change even a single value — unlike `yq`.
+
+_Is there an alternative to Kustomize and `yq` that is flexible and structured?_
+
+## Writing YAML with code
```
Comment (weibeld) [inline src/templating/content.md]: A more suitable title could be:

## Generating resource definitions with code

or

## Generating resource manifests with code

First, most libraries produce JSON and not YAML, and, second, what you actually want to create is a Kubernetes resource definition, not just arbitrary YAML or JSON.

File: src/templating/content.md [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+
+You could generate YAML programatically with code.
+
+And you don't even need to start from scratch.
+
+Kubernetes has several official libraries where you can create objects such as Deployments and Pod with code.
+
+> Please note that since Kubernetes uses an OpenAPI, you can autogenerate a library for your favourite language — if there isn't one already.
+
+As an example, let's have a look at how you can use Javascript to create a Pod and programatically change it.
```
Comment (weibeld) [inline src/templating/content.md]: ...how you can use JavaScript to generate a Kubernetes Pod definition.

File: src/templating/content.md [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+
```
Comment (weibeld) [inline src/templating/content.md]: First install the necessary Kubernetes JavaScript libraries:

```bash
npm init -y
npm install @kubernetes/client-node kubernetes-models
```

File: src/templating/content.md [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+```js|title=pod.js
```
Comment (weibeld) [inline src/templating/content.md]: Then, create the following JavaScript file:

File: src/templating/content.md [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+        ]
+      }
+    ]
+  },
+  "apiVersion": "v1",
+  "kind": "Pod"
+}
+```
+
+The output is a Pod, but it's in JSON and not YAML.
```
Comment (weibeld) [inline src/templating/content.md]: The output is a Pod definition in JSON.

File: src/templating/content.md [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+
+That shouldn't be a problem because:
+
+1. YAML is a superset of JSON. Any JSON file is also a valid YAML file
+1. The Kubernetes API receives the resource in JSON even if you write YAML files. `kubectl` serialises the YAML into JSON
```
Comment (weibeld) [inline src/templating/content.md]: Change this to something like:

>The Kubernetes API server accepts resource definitions in both JSON and YAML. YAML is often used when resource definitions are created manually, whereas JSON is often used when resource definitions are generated automatically.

File: src/templating/content.md [hunk 7]
```diff
@@ -0,0 +1,654 @@
...
+const pod = new Pod({
+  metadata: {
+    name: 'test-pod',
+  },
+  spec: {
+    containers: [
+      new Container({
+        name: 'test-container',
+        image: 'k8s.gcr.io/busybox',
+        env: [{ name: 'ENV', value: process.env.ENV || 'production' }],
```
Comment (weibeld) [inline src/templating/content.md]: Highlight line 12

File: src/templating/content.md [hunk 8]
```diff
@@ -0,0 +1,654 @@
...
+          }
+        ]
+      }
+    ]
+  },
+  "apiVersion": "v1",
+  "kind": "Pod"
+}
+```
+
```
Comment (weibeld) [inline src/templating/content.md]: You could save the above output in a file named `pod.json` and then create the Pod in the cluster with kubectl:

```bash
kubectl apply -f pod.json
```

File: src/templating/content.md [hunk 9]
```diff
@@ -0,0 +1,654 @@
...
+You could also skip `kubectl` all together and submit the JSON to your cluster.
```
Comment (weibeld) [inline src/templating/content.md]: ...and use JavaScript to submit the JSON to your cluster.

File: src/templating/content.md [hunk 10]
```diff
@@ -0,0 +1,654 @@
...
+
+Using the official Javascript library, you could have the following code:
```
Comment (weibeld) [inline src/templating/content.md]: You can do so by adding the following code to your existing script:

File: src/templating/content.md (hunk 3 above)
Comment (danielepolencic) [reply src/templating/content.md]: This is in the other repo. does it make sense to have it here as well?
I thought I could include only the relevant code and a link to the full code.
I think the focus is not on the completeness of the setup, but rather the intent of the code.
Looking for feedback here.

File: src/templating/content.md (hunk 6 above)
Comment (danielepolencic) [reply src/templating/content.md]: Is this true? I thought the serialisation to JSON happens in kubectl. I cannot do a curl request with some YAML to the API?

**Pull request merged**